sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/ContentDownloader.java

```diff
@@ -0,0 +1,298 @@
+
+        Lock progressLock = new ReentrantLock();
+        AtomicLong totalProgress = new AtomicLong(0);
+
+        Function<HttpRange, Mono<Response<Flux<ByteBuffer>>>> downloadFunc =
+            range -> downloadStreamingWithResponse(endpoint, range, context);
+
+        return downloadFirstChunk(parallelDownloadOptions, downloadFunc)
+            .flatMap(setupTuple2 -> {
+                long newCount = setupTuple2.getT1();
+                int numChunks = calculateNumBlocks(newCount, parallelDownloadOptions.getBlockSizeLong());
+
+                // In case it is an empty blob, this ensures we still actually perform a download operation.
+                numChunks = numChunks == 0 ? 1 : numChunks;
+
+                Response<Flux<ByteBuffer>> initialResponse = setupTuple2.getT2();
+                return Flux.range(0, numChunks)
+                    .flatMap(chunkNum -> downloadChunk(chunkNum, initialResponse,
+                        parallelDownloadOptions, newCount, downloadFunc,
+                        writer.apply(chunkNum).apply(progressLock).apply(totalProgress)))
+                    .then(Mono.just(new SimpleResponse<>(initialResponse, null)));
+            });
+    }
+
+    Mono<Response<Flux<ByteBuffer>>> downloadStreamingWithResponse(URI endpoint, HttpRange httpRange, Context context) {
+        Mono<HttpResponse> httpResponse = makeDownloadRequest(endpoint, httpRange, context);
+        return httpResponse.map(response -> {
+            Flux<ByteBuffer> result = getFluxStream(response, endpoint, httpRange, context);
+            return new SimpleResponse<>(response.getRequest(), response.getStatusCode(),
+                response.getHeaders(), result);
+        });
+    }
+
+    private Flux<ByteBuffer> getFluxStream(HttpResponse response, URI endpoint, HttpRange httpRange, Context context) {
+        return FluxUtil.createRetriableDownloadFlux(
+            () -> getResponseBody(response),
+            (Throwable throwable, Long aLong) -> {
+                HttpRange range;
+                if (httpRange != null) {
+                    range = new HttpRange(aLong + 1, httpRange.getLength() - aLong - 1);
+                } else {
+                    range = new HttpRange(aLong + 1);
+                }
+                Mono<HttpResponse> resumeResponse = makeDownloadRequest(endpoint, range, context);
+                return resumeResponse.map(this::getResponseBody).block();
+            },
+            Constants.ContentDownloader.MAX_RETRIES
+        );
+    }
+
+    private Flux<ByteBuffer> getResponseBody(HttpResponse response) {
+        switch (response.getStatusCode()) {
+            case 200:
+            case 206:
+                return response.getBody();
+            default:
+                throw logger.logExceptionAsError(
+                    new CommunicationErrorException(formatExceptionMessage(response), response)
```

Comment:
We have stopped the CommunicationErrorException from being exposed to the end users and instead created custom exceptions for each one of the different packages.

Here's an example on how we did that for the Identity client: https://github.com/Azure/azure-sdk-for-java/blob/a466b3d474290eacdbd1a7afcf211d431df7356f/sdk/communication/azure-communication-identity/src/main/java/com/azure/communication/identity/CommunicationIdentityAsyncClient.java#L251

And here's how we did it on the Phone Numbers Client: https://github.com/Azure/azure-sdk-for-java/blob/a466b3d474290eacdbd1a7afcf211d431df7356f/sdk/communication/azure-communication-phonenumbers/src/main/java/com/azure/communication/phonenumbers/PhoneNumbersAsyncClient.java#L230